src/Microsoft.ML.HalLearners/LRWithTrainingStatistics.cs

```diff
@@ -0,0 +1,346 @@
+            string labelColumn,
+            string weightColumn = null,
+            float l1Weight = Arguments.Defaults.L1Weight,
+            float l2Weight = Arguments.Defaults.L2Weight,
+            float optimizationTolerance = Arguments.Defaults.OptTol,
+            int memorySize = Arguments.Defaults.MemorySize,
+            bool enforceNoNegativity = Arguments.Defaults.EnforceNonNegativity,
+            Action<Arguments> advancedSettings = null)
+            : base(env, featureColumn, labelColumn, weightColumn,
+                  l1Weight, l2Weight, optimizationTolerance, memorySize, enforceNoNegativity, advancedSettings)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="LogisticRegression"/>
+        /// </summary>
+        internal LogisticRegressionWithStats(IHostEnvironment env, Arguments args)
+            : base(env, args)
+        {
+        }
+
+        protected override void ComputeTrainingStatistics(IChannel ch, FloatLabelCursor.Factory cursorFactory, float loss, int numParams)
+        {
+            Contracts.AssertValue(ch);
+            Contracts.AssertValue(cursorFactory);
+            Contracts.Assert(NumGoodRows > 0);
+            Contracts.Assert(WeightSum > 0);
+            Contracts.Assert(BiasCount == 1);
+            Contracts.Assert(loss >= 0);
+            Contracts.Assert(numParams >= BiasCount);
+            Contracts.Assert(CurrentWeights.IsDense);
+
+            ch.Info("Model trained with {0} training examples.", NumGoodRows);
+
+            // Compute deviance: start with loss function.
+            float deviance = (float)(2 * loss * WeightSum);
+
+            if (L2Weight > 0)
+            {
+                // Need to subtract L2 regularization loss.
+                // The bias term is not regularized.
+                var regLoss = VectorUtils.NormSquared(CurrentWeights.Values, 1, CurrentWeights.Length - 1) * L2Weight;
+                deviance -= regLoss;
+            }
+
+            if (L1Weight > 0)
+            {
+                // Need to subtract L1 regularization loss.
+                // The bias term is not regularized.
+                Double regLoss = 0;
+                VBufferUtils.ForEachDefined(ref CurrentWeights, (ind, value) => { if (ind >= BiasCount) regLoss += Math.Abs(value); });
+                deviance -= (float)regLoss * L1Weight * 2;
+            }
+
+            ch.Info("Residual Deviance: \t{0} (on {1} degrees of freedom)", deviance, Math.Max(NumGoodRows - numParams, 0));
+
+            // Compute null deviance, i.e., the deviance of null hypothesis.
+            // Cap the prior positive rate at 1e-15.
+            Double priorPosRate = PosWeight / WeightSum;
+            Contracts.Assert(0 <= priorPosRate && priorPosRate <= 1);
+            float nullDeviance = (priorPosRate <= 1e-15 || 1 - priorPosRate <= 1e-15) ?
+                0f : (float)(2 * WeightSum * MathUtils.Entropy(priorPosRate, true));
```

Comment:
if we go down this route, this will need to get refactored to avoid repetition. Doing it the fast way to see if it even passes the smell test :)
)